### PR TITLE
fix 714-plugin add netns validation reinforcement

### PR DESCRIPTION
**Pull request by Poor12**
fix-714: https://github.com/containernetworking/plugins/issues/714

### CODE REVIEW

**Comment by Poor12** [comment]
Is something like this? Need your review. :) @squeed

File: pkg/skel/skel.go [hunk 1]
```diff
@@ -190,6 +203,29 @@ func validateConfig(jsonBytes []byte) *types.Error {
...
+func getCurrentThreadNetNSPath() string {
+	// /proc/self/ns/net returns the namespace of the main thread, not
+	// of whatever thread this goroutine is running on.  Make sure we
+	// use the thread's net namespace since the thread is switching around
+	return fmt.Sprintf("/proc/%d/task/%d/ns/net", os.Getpid(), unix.Gettid())
+}
+
+func checkNetNS(nsPath string) (bool, *types.Error) {
+	ns, err := netns.GetFromPath(nsPath)
+	if err != nil {
```
Comment (squeed) [inline pkg/skel/skel.go:215]: We need to support CNI DEL for empty nsPath as well as already-deleted nsPath. So, paradoxically, I think we can return `true` here if `err != nil`.

File: pkg/skel/skel.go [hunk 2]
```diff
@@ -241,6 +277,17 @@ func (t *dispatcher) pluginMain(cmdAdd, cmdCheck, cmdDel func(_ *CmdArgs) error,
 		return types.NewError(types.ErrIncompatibleCNIVersion, "plugin version does not allow CHECK", "")
 	case "DEL":
 		err = t.checkVersionAndCall(cmdArgs, versionInfo, cmdDel)
+		if err != nil {
+			return err
+		}
+		if cmdArgs.NetnsOverride != "TRUE" {
```
Comment (dcbw) [inline pkg/skel/skel.go:283]: @Poor12 could we make this a case-insensitive comparison for "true", and also add a check for "1"?

File: pkg/skel/skel.go [hunk 3]
```diff
@@ -241,6 +277,17 @@ func (t *dispatcher) pluginMain(cmdAdd, cmdCheck, cmdDel func(_ *CmdArgs) error,
 		return types.NewError(types.ErrIncompatibleCNIVersion, "plugin version does not allow CHECK", "")
 	case "DEL":
 		err = t.checkVersionAndCall(cmdArgs, versionInfo, cmdDel)
+		if err != nil {
```
Comment (squeed) [inline pkg/skel/skel.go:280]: Should we check this for ADD as well? I don't see why not...

File: pkg/skel/skel.go (hunk 1 above)
Comment (Poor12) [reply pkg/skel/skel.go:215]: Do you mean we just pass the check and let plugins deal with empty plugins as well as already-deleted nsPath？

File: pkg/skel/skel.go (hunk 2 above)
Comment (Poor12) [reply pkg/skel/skel.go]: make it.

File: pkg/skel/skel.go (hunk 3 above)
Comment (Poor12) [reply pkg/skel/skel.go:302]: Add check for ADD.

**Comment by squeed** [comment]
@Poor12 surprise: this doesn't build on Windows :). We'll need to stub out the function with build flags.

**Comment by Poor12** [comment]
> @Poor12 surprise: this doesn't build on Windows :). We'll need to stub out the function with build flags.

I am not familiar with this. Could you do me a favor？Thanks a lot.

**Comment by dcbw** [comment]
> I am not familiar with this. Could you do me a favor？Thanks a lot.

```
 # github.com/vishvananda/netns
Error: C:\Users\runneradmin\go\pkg\mod\github.com\vishvananda\netns@v0.0.0-20210104183010-2eb08e3e575f\netns.go:28:13: undefined: unix.Stat_t
Error: C:\Users\runneradmin\go\pkg\mod\github.com\vishvananda\netns@v0.0.0-20210104183010-2eb08e3e575f\netns.go:29:12: undefined: unix.Fstat
Error: C:\Users\runneradmin\go\pkg\mod\github.com\vishvananda\netns@v0.0.0-20210104183010-2eb08e3e575f\netns.go:32:12: undefined: unix.Fstat
Error: C:\Users\runneradmin\go\pkg\mod\github.com\vishvananda\netns@v0.0.0-20210104183010-2eb08e3e575f\netns.go:43:8: undefined: unix.Stat_t
Error: C:\Users\runneradmin\go\pkg\mod\github.com\vishvananda\netns@v0.0.0-20210104183010-2eb08e3e575f\netns.go:44:12: undefined: unix.Fstat
Error: C:\Users\runneradmin\go\pkg\mod\github.com\vishvananda\netns@v0.0.0-20210104183010-2eb08e3e575f\netns.go:56:8: undefined: unix.Stat_t
Error: C:\Users\runneradmin\go\pkg\mod\github.com\vishvananda\netns@v0.0.0-20210104183010-2eb08e3e575f\netns.go:57:12: undefined: unix.Fstat
Error: C:\Users\runneradmin\go\pkg\mod\github.com\vishvananda\netns@v0.0.0-20210104183010-2eb08e3e575f\netns.go:71:12: undefined: unix.Close
```

That just means that because the netlink library and Unix stuff doesn't work on Windows, you need to move that code to a different file. Here's what you do.

1. Make a new file called netns_linux.go and put getCurrentNS(), getCurrentThreadNetNSPath(), and checkNetNS() into that.
2. Make a new file called netns_windows.go and stub out the checkNetNS() function there like:
```
func checkNetNS(nsPath string) (bool, *types.Error) {
    return false, nil
}
```

Files with _<platform> at the end are automatically only built on the platform in question. It's magic :) That way Linux builds will use the full implementation, and the Windows builds will use the stub implementation that just returns OK. @MikeZappa87 says that Windows doesn't have a good way (that we know of) to get the root compartment, so let's just ignore this on Windows for now.

**Comment by Poor12** [comment]
Thanks for informing this. I have not seen _windows files before. And Happy Labor Day :) @dcbw

File: pkg/ns/ns_windows.go [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ns
+
+func CheckNetNS(nsPath string) (bool, *types.Error) {
```
Comment (dcbw) [inline pkg/ns/ns_windows.go:17]: @Poor12 you need to import `"github.com/containernetworking/cni/pkg/types"` here to get types.Error.

Comment (Poor12) [reply pkg/ns/ns_windows.go:19]: Done

**Comment by Poor12** [comment]
Excuse me, several months ago I started a issue about https://github.com/containernetworking/plugins/issues/714 and have proposed a tentative solution but got very few replies lately.  Actually our productive environment has had this problem for a while and we really hope to fix this problem in the community in order to avoid the problem of upstream and downstream inconsistencies caused by our private modification.

But during several months waiting for a response from the community, I have several confusion about this problem.
1. Do we still need to fix this problem？Or we don't think it's a problem？
2. Is there a more reasonable solution for this problem？

cc @squeed @dcbw @mars1024 @matthewdupre @mccv1r0 @MikeZappa87 @jellonek
Really sorry if disturbing you.

**Pull request merged**